### ISSUE TITLE
TF-TRT vs UFF-TensorRT

### ISSUE THREAD

**Issue by PythonImageDeveloper** [NVIDIA/TensorRT#341]
I found that we can optimize the Tensorflow model in several ways. If I am mistaken, please tell me.

1- Using [TF-TRT](https://docs.nvidia.com/deeplearning/frameworks/tf-trt-user-guide/index.html), This API developer by tensorflow and integreted TensoRT to Tensorflow and this API called as :
`from tensorflow.python.compiler.tensorrt import trt_convert as trt` 
This API can be applied to any tensorflow models (new and old version models) without any converting error, because If this API don't support any new layers, don't consider these layers for TensorRT engines and these layers remain for Tensorflow engine and run on Tensorflow. right?

2- Using [TensorRT](https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html#overview), This API by developed by NVIDA and is independent of Tenorflow library (Not integrated to Tensorflow), and this API called as:
`import tensorrt as trt`
If we want to use this api, first, we must converting the tensorflow graph to UFF using uff-convertor and then parse the UFF graph to this API.
In this case, If the Tensorflow graph have unsupported layers we must use plugin or custom code for these layers, right?

3- I don't know, when we work with Tensorflow models, Why we use UFF converter then TensorRT, we can use directly TF-TRT API, right? If so, Are you tested the Tensorflow optimization model from these two method to get same performance? what's advantage of this UFF converter method?

I have some question about the two cases above:
4- I convert the ssd_mobilenet_v2 using two cases, In the case 1, I achieve slight improvement in speed but in the case 2, I achieve more improvement, why?
My opinion is that, In the case 1, The API only consider converting the precision (FP32 to FP16) and merging the possible layers together, But in the case 2, the graph is clean by UFF such as remove any redundant nodes like Asserts and Identity and then converted to tensorrt graph, right?

5- when we convert the trained model files like .ckpt and .meta, ... to frozen inference graph(.pb file), These layers don't remove from graph? only loss states and optimizer states , ... are removed?

**Comment by rmccorm4** [comment]
Hi @PythonImageDeveloper,

There is an additional route of using tf2onnx (https://github.com/onnx/tensorflow-onnx) to convert Tensorflow -> ONNX, and then using TensorRT API to convert from ONNX -> TensorRT, which is likely better supported than UFF -> TensorRT, since the UFF Parser will be deprecated in the future.

> I don't know, when we work with Tensorflow models, Why we use UFF converter then TensorRT, we can use directly TF-TRT API, right?

In general, I believe TensorRT will outperform TF-TRT if you can convert your entire model as is. However, when there are unsupported operations, TF-TRT will fall back to TF ops as you said, which makes it a bit easier to use. 

I don't think there are any up to date performance comparisons between pure TRT and TF-TRT. Additionally, you can create plugin layers for unsupported ops if only using TRT, and again, I don't know how that performance stacks up against using fallback TF ops with TF-TRT.

**Comment by jkjung-avt** [comment]
https://github.com/jkjung-avt/tensorrt_demos/issues/43#issuecomment-575739025
